ENH: Add strings_as_fixed_length parameter for df.to_records() (#18146)

qinghao1 opened the pull request:
This option changes DataFrame.to_records() dtype for string arrays
to 'Sx', where x is the length of the longest string, instead of 'O"

- [x] closes #18146
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

gfyoung commented on pandas/core/frame.py:
```diff
@@ -1351,6 +1352,9 @@ def to_records(self, index=True, convert_datetime64=None):
             DatetimeIndex.
+        strings_as_bytes : boolean, default False
+            Store strings as bytes (``S`` dtype) instead of Python objects
+            (``O`` dtype)
```
Add a `..versionadded` line right below your docstring description.

gfyoung commented on pandas/core/frame.py:
```diff
@@ -1436,7 +1458,17 @@ def to_records(self, index=True, convert_datetime64=None):
 
-        formats = [v.dtype for v in arrays]
+        if strings_as_bytes:
+            is_string = np.vectorize(lambda s: type(s) is str)
```
I would rather use `compat.string_types` for this check.

qinghao1 commented:
Thanks for the comments! Made the changes you requested.

jreback commented on pandas/core/frame.py:
```diff
@@ -1335,7 +1335,8 @@ def from_records(cls, data, index=None, exclude=None, columns=None,
         return cls(mgr)
 
-    def to_records(self, index=True, convert_datetime64=None):
+    def to_records(self, index=True, convert_datetime64=None,
```
so this is an extremely misleading parameter name. These are NOT bytes, rather they are fixed length strings. maybe ``strings_as_fixed_length``

jreback replied:
on second though call this ``stringlike_as_fixed_length``, to encompass string, unicode, bytes

jreback commented on pandas/tests/frame/test_convert_to.py:
```diff
@@ -186,6 +186,18 @@ def test_to_records_with_categorical(self):
+
+        # GH18146
+
+        df = DataFrame({'A': [1, 2], 'B': ['abc', 'defg']},
```
can you test with some mixed object dtypes that include strings, e.g. ``[1, '1']``, this should preserve object dtype.

jreback replied:
maybe just add another column to your test frame

qinghao1 commented:
@jreback Thanks for the feedback! I've made the changes.

jreback commented on pandas/core/frame.py:
```diff
@@ -1436,7 +1460,21 @@ def to_records(self, index=True, convert_datetime64=None):
 
-        formats = [v.dtype for v in arrays]
+        if strings_as_fixed_length:
+            def is_string(arr):
```
don't do this,  rather use

``lib.infer_dtype(arr) == 'string'`` (there is also a unicode variant, so I guess use 'U' for that)
for that matter if you actually have bytes it will return that as well.

Anything else would trigger 'O', note that this routine will correctly infer all dtypes

jreback replied:
e.g.

```
# this is just the public interface
In [3]: from pandas.api.types import infer_dtype

NameError: name 'ifner_dtype' is not defined

In [5]: infer_dtype(['foo', 'bar'])
Out[5]: 'string'

In [6]: infer_dtype([b'foo', b'bar'])
Out[6]: 'bytes'

# in PY2 this would be unicode
In [7]: infer_dtype([u'foo', u'bar'])
Out[7]: 'string'

In [8]: infer_dtype([u'foo', b'bar'])
Out[8]: 'mixed'
```

qinghao1 replied:
@jreback Sorry for the late reply. I think the original issue #18146 was meant for ASCII-encoded strings, and your comment has made me realize that this wouldn't work for strings with unicode characters. (Numpy throws an error like `UnicodeEncodeError: 'ascii' codec can't encode character '\u212b' in position 0: ordinal not in range(128)`).

I propose that this feature be implemented for `bytes` objects instead, which are still encoded as `O` dtype but can be more efficiently encoded as `S` dtype. And the option would be called `bytes_as_fixed_length`. What do you think?

jreback commented on pandas/tests/frame/test_convert_to.py:
```diff
@@ -186,6 +186,18 @@ def test_to_records_with_categorical(self):
+    def test_to_records_with_strings_as_fixed_length(self):
+
+        # GH18146
+
```
so test with strings, unicode, bytes

jreback commented:
can you rebase

jreback commented:
can you merge master

jreback commented:
@gfyoung would you be able to merge master and update this

gfyoung committed:
ENH: Allow fixed-length strings in df.to_records()

Adds parameter to allow string-like columns to be
cast as fixed-length string-like dtypes for more
efficient storage.

Closes gh-18146.

Originally authored by @qinghao1 but cleaned up
by @gfyoung to fix merge conflicts.

pep8speaks commented:
Hello @qinghao1! Thanks for updating the PR.

 - There are no PEP8 issues in the file [`pandas/core/frame.py`](https://github.com/qinghao1/pandas/blob/520e58ea152faf026eff322a00796721f1214259/pandas/core/frame.py) !

 - There are no PEP8 issues in the file [`pandas/tests/frame/test_convert_to.py`](https://github.com/qinghao1/pandas/blob/520e58ea152faf026eff322a00796721f1214259/pandas/tests/frame/test_convert_to.py) !

gfyoung commented:
@jreback : I rebased and addressed all of the comments, including leveraging `infer_dtype` more extensively in the enhancement + adding a lot more tests.

That being said, I'm not 100% certain if the design decisions made were the best.  I commented pretty extensively in the `frame.py` diff regarding what I did.

gfyoung commented on pandas/core/frame.py:
```diff
@@ -1569,7 +1596,47 @@ def to_records(self, index=True, convert_datetime64=None):
+                    # in Python 3.x. However, NumPy does
+                    # not have a fixed-length bytes dtype
+                    # and just uses string instead.
+                    symbol = "S"
```
Again, not sure this is best, but is the best fit I could find with how `numpy` and `Python 2.x` works.

gfyoung replied:
This is now moot.

gfyoung commented:
@jreback : The rebase worked, and all is green.  I left some comments for consideration during review.  PTAL

gfyoung commented:
@jreback : The rebase worked, and all is green.  I left some comments for consideration during review.  PTAL

gfyoung commented:
@jreback : Could you take a look at this?  Would need some feedback on some of the implementation choices I made here...

jreback commented on pandas/core/frame.py:
```diff
@@ -1476,7 +1476,8 @@ def from_records(cls, data, index=None, exclude=None, columns=None,
 
-    def to_records(self, index=True, convert_datetime64=None):
+    def to_records(self, index=True, convert_datetime64=None,
+                   stringlike_as_fixed_length=False):
```
maybe we should instead have a ``dtype=`` parameter? that accepts a dict?

gfyoung replied:
In place of the currently added parameter?  That would also work, as then we don’t have to deal with the messy dtype inference ourselves...

I’ll give that implementation a shot.

gfyoung replied:
In place of the currently added parameter?  That would also work, as then we don’t have to deal with the messy dtype inference ourselves...

I’ll give that implementation a shot.

gfyoung replied:
Because we can specify `dtype` for both columns and indices, added parameters for each.

Also considered just one giant `dtype` parameter to encompass both, but that felt a little cumbersome (e.g. you would always have nested dictionaries), especially people are probably used to generally specifying just column dtypes (e.g. `read_csv`).  Allowing the specification of `index` dtypes is a nice addition (and consistent with the original changes, which affected both indices and columns) nonetheless.

gfyoung committed:
Add dtype parameters instead of fix-string-like

The original parameter was causing a lot of acrobatics
with regards to string dtypes between 2.x and 3.x.

The new parameters simplify the internal logic and
pass the responsibility and motivation of memory
efficiency back to the users.

gfyoung commented on pandas/core/frame.py:
```diff
@@ -1569,7 +1596,47 @@ def to_records(self, index=True, convert_datetime64=None):
+                    # differentiate string from unicode
+                    # like NumPy arrays do, so we
+                    # specify unicode to be safe.
+                    symbol = "S" if compat.PY2 else "U"
```
This is now moot.

jreback commented on pandas/core/frame.py:
```diff
@@ -1598,6 +1612,18 @@ def to_records(self, index=True, convert_datetime64=None):
+
+        Data types can be specified for the columns:
+
+        >>> df.to_records(column_dtypes={"A": "int32"})
```
can you add an example which looks something like the OP, e.g. 

``df.to_records(index_dtypes="<S{}".format(df.index.str.len().max()))``

gfyoung replied:
Sure, done.

gfyoung commented:
@jreback : Addressed your comments, and all is green again.  PTAL.

jreback commented on pandas/core/frame.py:
```diff
@@ -1557,6 +1557,20 @@ def to_records(self, index=True, convert_datetime64=None):
+        column_dtypes : str, type, dict, default None
+            .. versionadded:: 0.24.0
+
+            If a string or type, the data type to store all columns. If
```
why is this anything but a dict of column_name -> dtype? what does the indices have to do with this?

gfyoung replied:
You can map column indices to dtypes e.g.

```
{0: "int32"}
```

maps `df.columns[0]` to the `int32` `dtype`.

This design was taken from `read_csv`, where we do something similar.

gfyoung replied:
> are you testing for a multi-index for the index? I think this is ok
multi-index on the columns? does a rec-array even work with this?

gfyoung replied:
Added tests for MI, both in the columns and indices.

jreback reviewed (changes requested):
are you testing for a multi-index for the index? I think this is ok
multi-index on the columns? does a rec-array even work with this?

gfyoung commented:
@jreback : Added more tests, and everything is finally all green again.  PTAL.

jreback commented:
@TomAugspurger thoughts on this.

TomAugspurger commented on pandas/core/frame.py:
```diff
@@ -1620,20 +1651,55 @@ def to_records(self, index=True, convert_datetime64=None):
+                dtype_mapping = column_dtypes
+                name = self.columns[index]
+
+            if isinstance(dtype_mapping, dict):
```
Don't hardcode dict here. Use `compat.Mapping`.

jreback replied:
actually use ``is_dict_like``

gfyoung replied:
Ah, good catch.  I’ll fix that.

gfyoung replied:
One thing though: `is_dict_like` does not guarantee existence of the `__contains__` method, which is what my implementation also relies on.

I could do some try-except logic to make it more generic, or I could make the `is_dict_like` check slightly stricter.

Thoughts?

jreback replied:
I would try and see if you can make is_dict_like stronger I think its ok here

jreback replied:
I really hate to use other things because they are completely unmaintanable in the codebase, hey just pick your favorite thing to check (dict, collections.Mapping), oh and you need something else, just reinvent it. 

This is the reason we have the is_* things.

gfyoung replied:
Okay, sounds good.  I was also leaning towards expanding the dict-like check but wanted to get some consensus.

I do understand the appeal of the collections checking, but as we already have the inference utilities, we should use them in the interest of maintainability.

TomAugspurger reviewed (commented):
+1 to the idea in general.

The implementation looks good aside from hardcoding dict.

My only quibble is with splitting `index_dtypes` and `column_dtypes`. Elsewhere in pandas (e.g. merge, groupby) we've allowed targeting index (levels) with keywords that were previously column-only. API-wise, I'd find that cleaner here. The downside would be cases where you want to apply a dtype all all the columns, and a different dtype to the index. In that case you would need to build a dict, rather than `(index_dtypes=int, column_dtypes=str).

gfyoung commented on pandas/core/frame.py:
```diff
@@ -1557,6 +1557,20 @@ def to_records(self, index=True, convert_datetime64=None):
 
             Whether to convert the index to datetime.datetime if it is a
             DatetimeIndex.
+        column_dtypes : str, type, dict, default None
```
Yep, that’s why I use `type` and not `numpy.dtype`.

TomAugspurger commented:
collections.abc.Mapping ensures that. Seems like exactly what we want here.

Messing with the various is_*_like methods makes me nervous.

On Fri, Dec 28, 2018 at 11:06 AM gfyoung <notifications@github.com> wrote:

> *@gfyoung* commented on this pull request.
> ------------------------------
>
> In pandas/core/frame.py
> <https://github.com/pandas-dev/pandas/pull/22229#discussion_r244365011>:
>
> > +
> +        index_len = len(index_names)
> +        formats = []
> +
> +        for i, v in enumerate(arrays):
> +            index = i
> +
> +            if index < index_len:
> +                dtype_mapping = index_dtypes
> +                name = index_names[index]
> +            else:
> +                index -= index_len
> +                dtype_mapping = column_dtypes
> +                name = self.columns[index]
> +
> +            if isinstance(dtype_mapping, dict):
>
> One thing though: is_dict_like does not guarantee existence of the
> __contains__ method, which is what my implementation also relies on.
>
> I could do some try-except logic to make it more generic, or I could make
> the is_dict_like check slightly stricter.
>
> Thoughts?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/22229#discussion_r244365011>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABQHIo4XYuy2E7pohfVAISp-Np7aMnRMks5u9k-jgaJpZM4VxcGz>
> .
>

TomAugspurger commented:
Unfortunately, isinstance checks on ABCs are relatively slow, so I don't
think it should be used everywhere.

On Fri, Dec 28, 2018 at 11:29 AM Tom Augspurger <tom.augspurger88@gmail.com>
wrote:

> FWIW, collections.abc.Mapping is *the* definition of dict-like in the
> Python world.
>
> On Fri, Dec 28, 2018 at 11:15 AM Jeff Reback <notifications@github.com>
> wrote:
>
>> *@jreback* commented on this pull request.
>> ------------------------------
>>
>> In pandas/core/frame.py
>> <https://github.com/pandas-dev/pandas/pull/22229#discussion_r244366683>:
>>
>> > +
>> +        index_len = len(index_names)
>> +        formats = []
>> +
>> +        for i, v in enumerate(arrays):
>> +            index = i
>> +
>> +            if index < index_len:
>> +                dtype_mapping = index_dtypes
>> +                name = index_names[index]
>> +            else:
>> +                index -= index_len
>> +                dtype_mapping = column_dtypes
>> +                name = self.columns[index]
>> +
>> +            if isinstance(dtype_mapping, dict):
>>
>> I really hate to use other things because they are completely
>> unmaintanable in the codebase, hey just pick your favorite thing to check
>> (dict, collections.Mapping), oh and you need something else, just reinvent
>> it.
>>
>> This is the reason we have the is_* things.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/pandas-dev/pandas/pull/22229#discussion_r244366683>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/ABQHIq0eEBMyVHCUczCsJStAAl1qhoU-ks5u9lGbgaJpZM4VxcGz>
>> .
>>
>

gfyoung commented:
@TomAugspurger : I actually commented about this earlier, but not seeing it here in the comments...weird...

I actually prefer to keep separate, in part because I suspect the use case in the wild  will largely focus on the columns, not the indices.  Thus, it makes sense to keep separate to facilitate this use case, while still making it possible to specify dtypes for the indices if so desired.

TomAugspurger commented:
Fair enough.

On Fri, Dec 28, 2018 at 12:38 PM gfyoung <notifications@github.com> wrote:

> @TomAugspurger <https://github.com/TomAugspurger> : I actually commented
> about this earlier, but not seeing it here in the comments...weird...
>
> I actually prefer to keep separate, in part because I suspect the use case
> in the wild will largely focus on the columns, not the indices. Thus, it
> makes sense to keep separate to facilitate this use case, while still
> making it possible to specify dtypes for the indices if so desired.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/22229#issuecomment-450406870>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABQHIkK0L3_zTj-1nfoSCKnfRWaOVmeyks5u9mUPgaJpZM4VxcGz>
> .
>

gfyoung committed:
MAINT: Use is_dict_like in to_records

More generic than checking whether our
mappings are instances of dict.

Expands is_dict_like check to include
whether it has a __contains__ method.

gfyoung commented:
@jreback @TomAugspurger : Managed to expand `is_dict_like` without issues, and I added a test in `to_records` to ensure that generic dict-like's are accepted.  PTAL.

jreback commented on pandas/core/dtypes/inference.py:
```diff
@@ -398,8 +398,11 @@ def is_dict_like(obj):
     >>> is_dict_like([1, 2, 3])
     False
     """
+    for attr in ("__getitem__", "keys", "__contains__"):
```
not a big deal, but can you expand the is_dict_like test to account for this?

gfyoung replied:
Sure thing.  I’ll do that and merge afterwards.

gfyoung replied:
Sure thing.  I’ll do that and merge afterwards.

gfyoung replied:
Sure thing.  I’ll do that and merge afterwards.

gfyoung committed:
TST: Add test for is_dict_like expanded def

gfyoung commented:
@jreback : Added a test for the expanded `is_dict_like`, which is all green.  PTAL.

jreback commented on pandas/core/frame.py:
```diff
@@ -1620,20 +1652,55 @@ def to_records(self, index=True, convert_datetime64=None):
             arrays = [self[c].get_values() for c in self.columns]
             names = lmap(compat.text_type, self.columns)
+            index_names = []
+
```
can you sprinkle a couple of comments in this to indicate what the flow is.

gfyoung replied:
Yep, done.

jreback commented on pandas/tests/dtypes/test_inference.py:
```diff
@@ -178,6 +178,33 @@ def test_is_dict_like_fails(ll):
+@pytest.mark.parametrize("has_keys", [True, False])
+@pytest.mark.parametrize("has_getitem", [True, False])
+@pytest.mark.parametrize("has_contains", [True, False])
+def test_is_dict_like_duct_type(has_keys, has_getitem, has_contains):
```
duct->duck

gfyoung replied:
Good catch 🙂

jreback reviewed (approved):
@gfyoung lgtm. just a couple of minor things. merge on green.

gfyoung committed:
MAINT: Address final comments

gfyoung commented:
Thanks for the help, @qinghao1 !
